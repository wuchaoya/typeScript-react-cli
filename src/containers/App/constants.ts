import { appThemes } from 'withStyles';

export const DEFAULT_THEMENAME: keyof typeof appThemes = 'light';
