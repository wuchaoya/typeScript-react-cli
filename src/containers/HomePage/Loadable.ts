import LoadingIndicator from '../../components/LoadingIndicator';
import { asyncComponent } from 'react-async-component';

export default asyncComponent({
  resolve: () => import(/* webpackChunkName: "home" */ './index'),
  LoadingComponent: LoadingIndicator
});
