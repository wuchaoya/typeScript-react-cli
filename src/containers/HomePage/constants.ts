export const LOAD_HITOKOTO = 'app/Home/CHANGE_USERNAME/LOAD_HITOKOTO';
export const LOAD_HITOKOTO_SUCCESS = 'app/Home/CHANGE_USERNAME/LOAD_HITOKOTO_SUCCESS';
export const LOAD_HITOKOTO_ERROR = 'app/Home/CHANGE_USERNAME/LOAD_HITOKOTO_ERROR';
