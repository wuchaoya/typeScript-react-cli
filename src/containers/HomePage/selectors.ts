import { IState } from 'Interfaces/state';
import { createSelector } from 'reselect';

export const selectHome = (state: IState) => state.get('home');

export const makeSelectHitokoto = () => createSelector(
  selectHome,
  (homeState) => homeState.get('hitokoto')
);
