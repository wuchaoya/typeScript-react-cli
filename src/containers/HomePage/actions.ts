import { LOAD_HITOKOTO, LOAD_HITOKOTO_SUCCESS, LOAD_HITOKOTO_ERROR } from 'containers/HomePage/constants';
import { createAction, createStandardAction } from 'typesafe-actions';

export const loadHitokoto = createAction(LOAD_HITOKOTO);

export const hitokotoLoaded = createStandardAction(LOAD_HITOKOTO_SUCCESS)<string>();

export const hitokotoLoadingError = createStandardAction(LOAD_HITOKOTO_ERROR)<object>();
