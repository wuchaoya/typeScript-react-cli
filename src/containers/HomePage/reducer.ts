import { fromJS } from 'immutable';
import { $Call, $Values } from 'utility-types';
import * as actions from './actions';
import { getType } from 'typesafe-actions';
export type HomeAction = $Call<$Values<typeof actions>>;

const initialState = fromJS({
  loading: false,
  error: false,
  hitokoto: undefined
});

export default function homeReducer(state = initialState, action: HomeAction) {
  switch (action.type) {
    case getType(actions.loadHitokoto):
      return state
        .set('loading', true)
        .set('error', false)
        .set('hitokoto', undefined);
    case getType(actions.hitokotoLoaded):
      return state
        .set('loading', false)
        .set('hitokoto', action.payload);
    case getType(actions.hitokotoLoadingError):
      return state
        .set('error', true)
        .set('loading', false);
    default:
      return state;
  }
}
