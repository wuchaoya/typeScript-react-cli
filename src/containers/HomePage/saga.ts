import { hitokotoLoaded, hitokotoLoadingError, loadHitokoto } from 'containers/HomePage/actions';
import { put, takeLatest, call } from 'redux-saga/effects';
import { LOAD_HITOKOTO } from 'containers/HomePage/constants';

export async function fetchHitokoto() {
  return fetch('https://sslapi.hitokoto.cn/').then(res => res.json());
}

export function* getHitokoto() {
  try {
    const data = yield call(fetchHitokoto);
    const hitokoto = data.hitokoto;
    yield put(hitokotoLoaded(hitokoto));
  } catch (err) {
    yield put(hitokotoLoadingError(err));
  }
}

export default function* hitokotoData() {
  yield takeLatest(LOAD_HITOKOTO, getHitokoto);
  yield put(loadHitokoto());
}
