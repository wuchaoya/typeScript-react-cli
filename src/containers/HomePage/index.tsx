
import { createSelector } from 'reselect';
import React from 'react';
import { Dispatch, compose } from 'redux';
import { connect } from 'react-redux';
import Helloworld from 'components/helloworld/index';
import { changeTheme } from 'containers/ThemeProvider/actions';
import { makeSelectThemeName } from 'containers/ThemeProvider/selectors';
import { appThemes } from 'withStyles';
import { loadHitokoto } from 'containers/HomePage/actions';
import reducer from './reducer';
import saga from './saga';

import injectReducer from 'utils/injectReducer';
import injectSaga from 'utils/injectSaga';
import { makeSelectHitokoto } from 'containers/HomePage/selectors';
import { $Call } from 'utility-types';
import { ONCE_TILL_UNMOUNT } from 'utils/constants';
import Button from '../../components/Button';

interface IHomePageProps {
}

const mapStateToProps = createSelector(
  makeSelectThemeName(),
  makeSelectHitokoto(),
  (theme, hitokoto) => ({ theme, hitokoto })
);

export const mapDispatchToProps = (dispatch: Dispatch) => ({
  changeTheme: (theme: string) => {dispatch(changeTheme(theme)); },
  onGetHitokoto: () => (dispatch(loadHitokoto()))
});

type stateProps = $Call<typeof mapStateToProps>;
type dispatchProps = $Call<typeof mapDispatchToProps>;

type Props = stateProps & IHomePageProps & dispatchProps;

export class HomePage extends React.PureComponent<Props, undefined> {

  handleThemeSelectChange: React.ReactEventHandler<HTMLSelectElement> = (e) => {
    this.props.changeTheme(e.currentTarget.value);
  }

  public render() {
    const appThemesItems = Object.keys(appThemes).map((theme) => {
      return <option key={theme} value={theme}>{theme}</option>;
    });
    return (
      <div>
        <Helloworld />
        <select
          name='Theme'
          value={this.props.theme}
          onChange={this.handleThemeSelectChange}
        >
          {appThemesItems}
        </select>
        <Button buttonText={'呵呵'} />
        <p>{this.props.hitokoto}</p>
      </div>
    );
  }
}

// tslint:disable-next-line:max-line-length
const withConnect = connect<stateProps, dispatchProps, IHomePageProps>(mapStateToProps, mapDispatchToProps);

const withReducer = injectReducer({ key: 'home', reducer });
const withSaga = injectSaga({ key: 'home', saga, mode: ONCE_TILL_UNMOUNT });

export default compose(
  withReducer,
  withSaga,
  withConnect
)(HomePage);
