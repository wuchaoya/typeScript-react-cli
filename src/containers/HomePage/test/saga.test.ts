import hitokotoData, {getHitokoto, fetchHitokoto} from 'containers/HomePage/saga';
import {put, takeLatest, call} from 'redux-saga/effects';
import {hitokotoLoaded, hitokotoLoadingError, loadHitokoto} from 'containers/HomePage/actions';
import {LOAD_HITOKOTO} from 'containers/HomePage/constants';

describe('getHitokoto Saga', () => {
  let getHitokotoGenerator: IterableIterator<any>;

  it('should call the fetchHitokoto function', () => {
    getHitokotoGenerator = getHitokoto();

    const callfetchDescriptor = getHitokotoGenerator.next().value;
    expect(callfetchDescriptor).toEqual(call(fetchHitokoto));
  });

  describe('after requests', () => {
    beforeEach(() => {
      getHitokotoGenerator = getHitokoto();

      const callfetchDescriptor = getHitokotoGenerator.next().value;
      expect(callfetchDescriptor).toMatchSnapshot();
    });

    it('should dispatch the hitokotoLoaded action if it requests the data successfully', () => {
      const hitokoto = '123';
      const putDescriptor = getHitokotoGenerator.next({hitokoto}).value;
      expect(putDescriptor).toEqual(put(hitokotoLoaded(hitokoto)));
    });

    it('should call the hitokotoLoadingError action if the response errors', () => {
      const response = new Error('Some error');
      const putDescriptor = getHitokotoGenerator.throw(response).value;
      expect(putDescriptor).toEqual(put(hitokotoLoadingError(response)));
    });
  });
});

describe('hitokotoData Saga', () => {
  let hitokotoDataSaga: IterableIterator<any>;

  beforeEach(() => {
    hitokotoDataSaga = hitokotoData();
  });

  it('should start task to watch for LOAD_HITOKOTO action', () => {
    const takeLatestDescriptor = hitokotoDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_HITOKOTO, getHitokoto));
  });

  it('should put LOAD_HITOKOTO action', () => {
    hitokotoDataSaga.next();
    const putDescriptor = hitokotoDataSaga.next().value;
    expect(putDescriptor).toEqual(put(loadHitokoto()));
  });
});
