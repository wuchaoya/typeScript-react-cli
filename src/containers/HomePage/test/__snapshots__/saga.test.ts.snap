// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getHitokoto Saga after requests should call the hitokotoLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getHitokoto Saga after requests should dispatch the hitokotoLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;
